File: python/src/azoth/eos/reference/soreide_whitson_phase.py
"""``eos.soreide_whitson_phase`` - the phase state of a Soreide-Whitson fluid.

Spec: ``specs/models/eos/soreide_whitson_phase.toml``. A *direct* model: the root is the
cubic's, chosen by ordering, so no iteration and no algorithm block.

This is the pure-Python reference: a second, independent expression of the same physics as
the Rust kernel, held to it by ``test_cross_impl.py``. What is here is the assembly - the
fluid is ``soreide_whitson_mixture_of``'s, the reduced parameters and the phase state are
``_mixture_state``'s, and the one line this model adds is that the interaction matrix is
resolved **at the phase's composition** before either is evaluated.
"""

from __future__ import annotations

from collections.abc import Sequence

from azoth import _models_gen
from azoth.core.errors import InvalidInputError
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import SoreideWhitsonPhaseResult
from azoth.core.units import Q, input_to_si
from azoth.core.warnings import Warning
from azoth.eos.components import soreide_whitson_mixture_of
from azoth.eos.reference._mixture_state import phase_state, reduced_parameters

MODEL_ID = "eos.soreide_whitson_phase"

#: The roots this model has, as the databank's cubic root convention names them.
_PHASES = {"liquid": True, "vapour": False, "vapor": False}


def soreide_whitson_phase(
    components: Sequence[str],
    T: Q,
    P: Q,
    x: Sequence[float],
    salinity: Q,
    compressed_phase: str,
) -> SoreideWhitsonPhaseResult:
    """One Soreide-Whitson phase's state at a temperature, pressure and composition.

    Args:
        components: the fluid's substances, by name. Each name decides the role it takes
            in the aqueous correlation.
        T: absolute temperature.
        P: absolute pressure.
        x: the phase's mole fractions, checked rather than renormalised.
        salinity: the equivalent-NaCl molality of the brine, in mol per kg of water.
        compressed_phase: ``"liquid"`` or ``"vapour"``.

    Returns:
        The compressibility factor at the chosen root and the fugacity coefficients.

    Raises:
        InvalidInputError: if ``x`` is not a composition, ``compressed_phase`` is neither
            spelling, a name is not in the databank, an ion is named, or ``salinity`` is
            negative.
        PropertyUnavailableError: if a component carries no heat-capacity coefficients.
        OutOfRangeError: if ``T`` or ``P`` is not positive, or no volume root exists.

    See :func:`azoth.eos.soreide_whitson_phase`.
    """
    spec = _models_gen.model(MODEL_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []

    t_si = input_to_si(spec, "T", T)
    p_si = input_to_si(spec, "P", P)
    salinity_si = input_to_si(spec, "salinity", salinity)
    apply_checks(
        checks.on_input, {"T": t_si, "P": p_si, "salinity": salinity_si}.get, warnings
    )

    if len(components) != len(x):
        raise InvalidInputError(
            "x", f"{len(components)} components but {len(x)} fractions; the two must match"
        )
    total = sum(x)
    if abs(total - 1.0) > 1.0e-9:
        raise InvalidInputError(
            "x",
            f"the mole fractions sum to {total}, not to one. Renormalising them here "
            "would make a composition error invisible in every number downstream, so "
            "it is refused instead",
        )
    if compressed_phase not in _PHASES:
        raise InvalidInputError(
            "compressed_phase",
            f"is {compressed_phase!r}; the roots this model has are {sorted(_PHASES)}",
        )

    fluid, _ = soreide_whitson_mixture_of(list(components), salinity_si)
    reduced = reduced_parameters(fluid, t_si, p_si)
    warnings.extend(reduced.warnings)
    # `phase_kij` and not `fluid.kij`: the rule replaces the water-gas entries of a
    # water-rich phase, so the matrix is a function of the composition this state is at.
    state = phase_state(
        reduced, fluid.phase_kij(reduced, list(x)), list(x), liquid=_PHASES[compressed_phase]
    )

    return SoreideWhitsonPhaseResult(
        z_factor=state.z,
        ln_phi=tuple(state.ln_phi),
        warnings=tuple(warnings),
    )


__all__ = ["soreide_whitson_phase"]

File: crates/azoth-eos/src/soreide_whitson_phase.rs
//! `eos.soreide_whitson_phase` - the phase state of a Soreide-Whitson fluid.
//!
//! Spec: `specs/models/eos/soreide_whitson_phase.toml`. A *direct* model: the root is the
//! cubic's, chosen by ordering, so no iteration and no `algorithm` block.
//!
//! NeqSim's `SystemSoreideWhitson`, whose distance from Peng-Robinson 1978 is exactly two
//! things:
//!
//! ```text
//! kij(i, water) = the salinity correlation, in a water-rich phase
//! alpha(water)  = AttractiveTermSoreideWhitson, salinity-dependent
//! ```
//!
//! Everything else - every other pair, every other component, every phase whose water is
//! below a mole fraction of `0.8` - is PR78. The two are independent paths and this model
//! is where they meet: [`crate::soreide_whitson_alpha`] and
//! [`crate::mixing_rule::MixingRule::phase_kij`] are the pieces, and this is the assembly.
//!
//! # What this model does not have
//!
//! **No flash.** `eos.pt_flash`'s Jacobian is `d ln phi / d n`, which
//! [`crate::mixture::Mixture::phase_derivatives`] refuses for this rule because its
//! interaction matrix moves with the composition being differentiated against. The
//! boundary is the one `MixingRule::HuronVidal` and `MixingRule::WongSandler` sit behind.
//!
//! **And the salinity is the caller's.** NeqSim's `calcSalinity` recomputes the phase's
//! molality from the flashed aqueous phase and pushes it into the alpha through
//! `setSalinityFromPhase` - a call its own `getClass().getName()` test never reaches,
//! which is the defect reported as the alpha issue. A caller here states the brine, and
//! the model states what it read.

use azoth_core::Result;
use azoth_core::units::{Pressure, ThermodynamicTemperature};

use crate::model_gen;
use crate::results::SoreideWhitsonPhaseResult;

/// One Soreide-Whitson phase's state at a temperature, pressure and composition.
///
/// # Errors
/// * [`azoth_core::AzothError::InvalidInput`] if `x` is not one entry per component or is
///   not a composition, a name is not in the databank, an ion is named, the salinity is
///   negative, or `compressed_phase` is neither root's spelling.
/// * [`azoth_core::AzothError::PropertyUnavailable`] if a component carries no
///   heat-capacity coefficients.
/// * [`azoth_core::AzothError::OutOfRange`] if `T` or `P` is not positive, or no volume
///   root exists above the mixture's covolume.
#[allow(non_snake_case)] // `T` and `P` are the symbols in the chemistry
pub fn soreide_whitson_phase(
    components: &[String],
    T: ThermodynamicTemperature,
    P: Pressure,
    x: &[f64],
    salinity: f64,
    compressed_phase: &str,
) -> Result<SoreideWhitsonPhaseResult> {
    let spec = &model_gen::SOREIDE_WHITSON_PHASE_SPEC;
    let mut warnings = Vec::new();
    azoth_core::apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "T" => Some(T.value),
            "P" => Some(P.value),
            "salinity" => Some(salinity),
            _ => None,
        },
        &mut warnings,
    )?;

    let n = components.len();
    if x.len() != n {
        return Err(azoth_core::AzothError::invalid_input(
            "x",
            format!("a mixture of {n} components has {} mole fractions", x.len()),
        ));
    }
    let sum: f64 = x.iter().sum();
    if (sum - 1.0).abs() > 1.0e-09 {
        return Err(azoth_core::AzothError::invalid_input(
            "x",
            format!(
                "the mole fractions sum to {sum}, not to one. Renormalising them here \
                 would make a composition error invisible in every number downstream, so \
                 it is refused instead"
            ),
        ));
    }

    let side = crate::cpa_phase::side_of(compressed_phase)?;
    let names: Vec<&str> = components.iter().map(String::as_str).collect();
    let (mixture, _) = crate::databank::soreide_whitson_mixture_of(&names, salinity, None)?;
    let reduced = mixture.reduced_parameters(T, P)?;
    warnings.extend(reduced.warnings.iter().cloned());
    let state = mixture.phase_state(&reduced, x, side)?;

    Ok(SoreideWhitsonPhaseResult {
        z_factor: state.z,
        ln_phi: state.ln_phi,
        warnings,
    })
}
